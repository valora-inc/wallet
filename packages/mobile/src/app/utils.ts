import { DappRequestOrigin } from 'src/analytics/types'
import { ActiveDapp } from 'src/app/reducers'

// Assume that if we have an active dapp, any WC or DappKit request comes from the in-app webview.
// Note that this may be incorrect if the user interacts with another dapp while the dapp webview is open.
// I thought about comparing the request url to the active dapp url, but that could fail too if the dapp url we have is different.
// For instance the dapp-list could contain a "example.com", while the request is "dapp.something.com".
export function getDappRequestOrigin(activeDapp: ActiveDapp | null) {
  if (activeDapp) {
    return DappRequestOrigin.InAppWebView
  }

  return DappRequestOrigin.External
}
