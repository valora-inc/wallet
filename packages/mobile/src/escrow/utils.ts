import { ensureLeading0x } from '@celo/base'
import { generateDeterministicInviteCode } from '@celo/utils/lib/account'
import { publicKeyToAddress } from '@celo/utils/lib/address'
import { EscrowedPayment } from 'src/escrow/actions'
import { identifierToE164NumberSelector } from 'src/identity/selectors'
import { ContactRecipient } from 'src/recipients/recipient'
import useSelector from 'src/redux/useSelector'

export const generateEscrowPaymentIdAndPk = (
  recipientPhoneHash: string,
  recipientPepper: string,
  addressIndex: number = 0
) => {
  const { publicKey, privateKey } = generateDeterministicInviteCode(
    recipientPhoneHash,
    recipientPepper,
    addressIndex
  )
  return { paymentId: publicKeyToAddress(publicKey), privateKey: ensureLeading0x(privateKey) }
}

export const generateUniquePaymentId = (
  existingPaymentIds: string[],
  phoneHash: string,
  pepper: string
) => {
  const paymentIdSet: Set<string> = new Set(existingPaymentIds)

  // Using an upper bound of 100 to be sure this doesn't run forever given
  // the realistic amount of pending escrow txs is far less than this
  for (let i = 0; i < 100; i += 1) {
    const { paymentId } = generateEscrowPaymentIdAndPk(phoneHash, pepper, i)
    if (!paymentIdSet.has(paymentId)) {
      return paymentId
    }
  }
}

export const useEscrowPaymentRecipient = (payment: EscrowedPayment): ContactRecipient => {
  const { recipientPhone, recipientIdentifier } = payment

  const identifierToE164Number = useSelector(identifierToE164NumberSelector)
  const recipientCache = useSelector((state) => state.recipients.phoneRecipientCache)
  const phoneNumber = identifierToE164Number[recipientIdentifier] ?? recipientPhone
  const recipient = recipientCache[phoneNumber] ?? {
    name: phoneNumber,
    e164PhoneNumber: phoneNumber,
    contactId: '',
  }

  return recipient
}
