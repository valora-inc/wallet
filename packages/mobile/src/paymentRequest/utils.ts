import { hexToBuffer } from '@celo/utils/lib/address'
import { decryptComment, encryptComment } from '@celo/utils/lib/commentEncryption'
import { isE164Number } from '@celo/utils/lib/phoneNumbers'
import { call } from 'redux-saga/effects'
import { MAX_COMMENT_LENGTH } from 'src/config'
import { features } from 'src/flags'
import i18n from 'src/i18n'
import { PaymentRequest, WriteablePaymentRequest } from 'src/paymentRequest/types'
import Logger from 'src/utils/Logger'
import { doFetchDataEncryptionKey } from 'src/web3/dataEncryptionKey'

const TAG = 'paymentRequest/utils'

// Encrypt sensitive data in the payment request using the recipient and sender DEK
export function* encryptPaymentRequest(paymentRequest: WriteablePaymentRequest) {
  Logger.debug(`${TAG}@encryptPaymentRequest`, 'Encrypting payment request')

  const fromKey: Buffer | null = yield call(
    doFetchDataEncryptionKey,
    paymentRequest.requesterAddress
  )
  if (!fromKey) {
    Logger.debug(`${TAG}@encryptPaymentRequest`, 'No sender key found, skipping encryption')
    return sanitizePaymentRequest(paymentRequest)
  }

  const toKey: Buffer | null = yield call(doFetchDataEncryptionKey, paymentRequest.requesteeAddress)
  if (!toKey) {
    Logger.debug(`${TAG}@encryptPaymentRequest`, 'No recipient key found, skipping encryption')
    return sanitizePaymentRequest(paymentRequest)
  }

  const encryptedPaymentRequest: WriteablePaymentRequest = {
    ...paymentRequest,
  }

  if (paymentRequest.requesterE164Number) {
    // Using the same util as for comment encryption to encrypt the phone number
    // TODO: Consider renaming this util for clarity
    const { comment: encryptedE164Number, success } = encryptComment(
      paymentRequest.requesterE164Number,
      toKey,
      fromKey
    )

    // We intentionally exclude the phone number if we can't encrypt it
    // The request still contains an address which may still map to a contact
    // If the recipient has seen it before.
    encryptedPaymentRequest.requesterE164Number = success ? encryptedE164Number : undefined
  }

  const comment = paymentRequest.comment
  if (comment && features.USE_COMMENT_ENCRYPTION) {
    const { comment: encryptedComment, success } = encryptComment(comment, toKey, fromKey)
    encryptedPaymentRequest.comment = success ? encryptedComment : comment
  }

  return encryptedPaymentRequest
}

// Decrypt sensitive data in the payment request using the user's DEK
export function decryptPaymentRequest(
  paymentRequest: PaymentRequest,
  dataEncryptionKey: string | null,
  isOutgoingRequest: boolean
) {
  Logger.debug(`${TAG}@decryptPaymentRequest`, 'Decrypting payment request')

  if (!dataEncryptionKey) {
    Logger.error(`${TAG}@decryptPaymentRequest`, 'Missing DEK, should never happen.')
    return paymentRequest
  }
  const dekBuffer = hexToBuffer(dataEncryptionKey)

  const decryptedPaymentRequest: PaymentRequest = {
    ...paymentRequest,
  }

  const requesterE164Number = paymentRequest.requesterE164Number
  if (requesterE164Number) {
    const { comment: decryptedRequesterE164Number, success } = decryptComment(
      requesterE164Number,
      dekBuffer,
      isOutgoingRequest
    )
    if (success) {
      decryptedPaymentRequest.requesterE164Number = decryptedRequesterE164Number
    } else if (isE164Number(requesterE164Number)) {
      Logger.warn(
        `${TAG}@decryptPaymentRequest`,
        'Decrypting requesterE164Number failed, using raw number'
      )
      decryptedPaymentRequest.requesterE164Number = requesterE164Number
    } else {
      Logger.warn(
        `${TAG}@decryptPaymentRequest`,
        'requesterE164Number appears to be ciphertext, excluding it'
      )
      decryptedPaymentRequest.requesterE164Number = undefined
    }
  }

  const comment = paymentRequest.comment
  if (comment && features.USE_COMMENT_ENCRYPTION) {
    const { comment: decryptedComment, success } = decryptComment(
      comment,
      dekBuffer,
      isOutgoingRequest
    )
    if (success) {
      decryptedPaymentRequest.comment = decryptedComment
    } else if (comment.length <= MAX_COMMENT_LENGTH) {
      Logger.warn(`${TAG}@decryptPaymentRequest`, 'Decrypting comment failed, using raw comment')
      decryptedPaymentRequest.comment = comment
    } else {
      Logger.warn(
        `${TAG}@decryptPaymentRequest`,
        'Comment appears to be ciphertext, hiding comment'
      )
      decryptedPaymentRequest.comment = i18n.t('commentUnavailable')
    }
  }

  return decryptedPaymentRequest
}

// For cases when the request can't be encrypted, remove sensitive PII
function sanitizePaymentRequest(paymentRequest: WriteablePaymentRequest): WriteablePaymentRequest {
  return {
    ...paymentRequest,
    requesterE164Number: undefined,
  }
}
