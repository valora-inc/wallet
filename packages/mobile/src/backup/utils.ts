import { MnemonicLanguages, normalizeMnemonic } from '@celo/utils/lib/account'
import CryptoJS from 'crypto-js'
import { useAsync } from 'react-async-hook'
import { useDispatch, useSelector } from 'react-redux'
import { showError } from 'src/alert/actions'
import { OnboardingEvents } from 'src/analytics/Events'
import ValoraAnalytics from 'src/analytics/ValoraAnalytics'
import { ErrorMessages } from 'src/app/ErrorMessages'
import { getPassword } from 'src/pincode/authentication'
import { removeStoredItem, retrieveStoredItem, storeItem } from 'src/storage/keychain'
import Logger from 'src/utils/Logger'
import { currentAccountSelector } from 'src/web3/selectors'

const TAG = 'Backup/utils'

export const DAYS_TO_BACKUP = 1
export const DAYS_TO_DELAY = 1 / 24 // 1 hour delay
export const MNEMONIC_SPLITTER = 'celo'

export const MNEMONIC_STORAGE_KEY = 'mnemonic'

export function getMnemonicLanguage(language: string | null) {
  switch (language?.slice(0, 2)) {
    case 'es': {
      return MnemonicLanguages.spanish
    }
    case 'pt': {
      return MnemonicLanguages.portuguese
    }
    default: {
      return MnemonicLanguages.english
    }
  }
}

export async function storeMnemonic(mnemonic: string, account: string | null, password?: string) {
  if (!account) {
    throw new Error('Account not yet initialized')
  }
  const passwordToUse = password ?? (await getPassword(account))
  const encryptedMnemonic = await encryptMnemonic(mnemonic, passwordToUse)
  return storeItem({ key: MNEMONIC_STORAGE_KEY, value: encryptedMnemonic })
}

export async function clearStoredMnemonic() {
  await removeStoredItem(MNEMONIC_STORAGE_KEY)
}

export async function getStoredMnemonic(
  account: string | null,
  password?: string
): Promise<string | null> {
  try {
    if (!account) {
      throw new Error('Account not yet initialized')
    }

    Logger.debug(TAG, 'Checking keystore for mnemonic')
    const encryptedMnemonic = await retrieveStoredItem(MNEMONIC_STORAGE_KEY)
    if (!encryptedMnemonic) {
      throw new Error('No mnemonic found in storage')
    }

    const passwordToUse = password ?? (await getPassword(account))
    return decryptMnemonic(encryptedMnemonic, passwordToUse)
  } catch (error) {
    Logger.error(TAG, 'Failed to retrieve mnemonic', error)
    return null
  }
}

export function onGetMnemonicFail(viewError: (error: ErrorMessages) => void, context?: string) {
  viewError(ErrorMessages.FAILED_FETCH_MNEMONIC)
  ValoraAnalytics.track(OnboardingEvents.backup_error, {
    error: 'Failed to retrieve Account Key',
    context,
  })
}

export function useAccountKey(): string | null {
  const dispatch = useDispatch()
  const account = useSelector(currentAccountSelector)
  const asyncAccountKey = useAsync(getStoredMnemonic, [account])

  if (!asyncAccountKey || asyncAccountKey.error) {
    onGetMnemonicFail((error) => dispatch(showError(error)), 'useAccountKey')
  }

  return asyncAccountKey.result || null
}

export function countMnemonicWords(phrase: string): number {
  return [...phrase.trim().split(/\s+/)].length
}

// Because of a RN bug, we can't fully clean the text as the user types
// https://github.com/facebook/react-native/issues/11068
export function formatBackupPhraseOnEdit(phrase: string) {
  return phrase.replace(/\s+/gm, ' ')
}

export function formatBackupPhraseOnSubmit(phrase: string) {
  return normalizeMnemonic(phrase)
}

function isValidMnemonic(phrase: string, length: number) {
  return !!phrase && countMnemonicWords(formatBackupPhraseOnEdit(phrase)) === length
}

export function isValidBackupPhrase(phrase: string) {
  return isValidMnemonic(phrase, 24)
}

export async function encryptMnemonic(phrase: string, password: string) {
  return CryptoJS.AES.encrypt(phrase, password).toString()
}

export async function decryptMnemonic(encryptedMnemonic: string, password: string) {
  const bytes = CryptoJS.AES.decrypt(encryptedMnemonic, password)
  return bytes.toString(CryptoJS.enc.Utf8)
}
