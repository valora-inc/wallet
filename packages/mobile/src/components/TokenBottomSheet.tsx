import Touchable from '@celo/react-components/components/Touchable'
import colors from '@celo/react-components/styles/colors'
import fontStyles from '@celo/react-components/styles/fonts'
import { Spacing } from '@celo/react-components/styles/styles'
import React, { useState } from 'react'
import { useTranslation } from 'react-i18next'
import { Dimensions, Image, LayoutChangeEvent, StyleSheet, Text, View } from 'react-native'
import { TouchableWithoutFeedback } from 'react-native-gesture-handler'
import Animated, { useAnimatedStyle, useSharedValue } from 'react-native-reanimated'
import { useSafeAreaInsets } from 'react-native-safe-area-context'
import { SendEvents } from 'src/analytics/Events'
import ValoraAnalytics from 'src/analytics/ValoraAnalytics'
import TokenDisplay from 'src/components/TokenDisplay'
import { useShowOrHideAnimation } from 'src/components/useShowOrHideAnimation'
import useSelector from 'src/redux/useSelector'
import { TokenBalance } from 'src/tokens/reducer'
import { coreTokensSelector, tokensWithTokenBalanceSelector } from 'src/tokens/selectors'
import { sortByUsdBalance } from 'src/tokens/utils'

export enum TokenPickerOrigin {
  Send = 'Send',
  SendConfirmation = 'SendConfirmation',
  Exchange = 'Exchange',
}

interface Props {
  isVisible: boolean
  origin: TokenPickerOrigin
  onTokenSelected: (tokenAddress: string) => void
  onClose: () => void
  isOutgoingPaymentRequest?: boolean
}

const MIN_EMPTY_SPACE = 100

function TokenOption({ tokenInfo, onPress }: { tokenInfo: TokenBalance; onPress: () => void }) {
  return (
    <Touchable onPress={onPress} testID={`${tokenInfo.symbol}Touchable`}>
      <View style={styles.tokenOptionContainer}>
        <Image source={{ uri: tokenInfo.imageUrl }} style={styles.tokenImage} />
        <View style={styles.tokenNameContainer}>
          <Text style={styles.localBalance}>{tokenInfo.symbol}</Text>
          <Text style={styles.currencyBalance}>{tokenInfo.name}</Text>
        </View>
        <View style={styles.tokenBalanceContainer}>
          <TokenDisplay
            style={styles.localBalance}
            amount={tokenInfo.balance}
            tokenAddress={tokenInfo.address}
            showLocalAmount={true}
            testID={`Local${tokenInfo.symbol}Balance`}
          />
          <TokenDisplay
            style={styles.currencyBalance}
            amount={tokenInfo.balance}
            tokenAddress={tokenInfo.address}
            showLocalAmount={false}
            testID={`${tokenInfo.symbol}Balance`}
          />
        </View>
      </View>
    </Touchable>
  )
}
// TODO: In the exchange flow or when requesting a payment, only show CELO & stable tokens.
function TokenBottomSheet({
  isVisible,
  origin,
  onTokenSelected,
  onClose,
  isOutgoingPaymentRequest,
}: Props) {
  const [showingOptions, setOptionsVisible] = useState(isVisible)
  const [pickerHeight, setPickerHeight] = useState(0)

  const tokens = useSelector(tokensWithTokenBalanceSelector)
  const coreTokens = useSelector(coreTokensSelector)
  const tokenList = (isOutgoingPaymentRequest ? coreTokens : tokens).sort(sortByUsdBalance)

  const { t } = useTranslation()

  const onTokenPressed = (tokenAddress: string) => () => {
    ValoraAnalytics.track(SendEvents.token_selected, {
      origin,
      tokenAddress,
    })
    onTokenSelected(tokenAddress)
  }

  const safeAreaInsets = useSafeAreaInsets()

  const progress = useSharedValue(0)
  const animatedPickerPosition = useAnimatedStyle(() => ({
    transform: [{ translateY: (1 - progress.value) * pickerHeight }],
  }))
  const animatedOpacity = useAnimatedStyle(() => ({
    opacity: 0.5 * progress.value,
  }))

  useShowOrHideAnimation(
    progress,
    isVisible,
    () => setOptionsVisible(true),
    () => setOptionsVisible(false)
  )

  if (!showingOptions) {
    return null
  }

  const onLayout = (event: LayoutChangeEvent) => {
    const { height } = event.nativeEvent.layout
    setPickerHeight(height)
  }

  const maxHeight = Dimensions.get('window').height - MIN_EMPTY_SPACE
  const paddingBottom = Math.max(safeAreaInsets.bottom, Spacing.Thick24)

  return (
    <View style={styles.container} testID="TokenBottomSheetContainer">
      <Animated.View style={[styles.background, animatedOpacity]}>
        <TouchableWithoutFeedback
          style={styles.backgroundTouchable}
          onPress={onClose}
          testID={'BackgroundTouchable'}
        />
      </Animated.View>
      <Animated.ScrollView
        style={[styles.contentContainer, { paddingBottom, maxHeight }, animatedPickerPosition]}
        contentContainerStyle={pickerHeight >= maxHeight ? styles.fullHeightScrollView : undefined}
        onLayout={onLayout}
      >
        <Text style={styles.title}>{t('selectToken')}</Text>
        {tokenList.map((tokenInfo, index) => {
          return (
            <React.Fragment key={`token-${tokenInfo.address}`}>
              {index > 0 && <View style={styles.separator} />}
              <TokenOption tokenInfo={tokenInfo} onPress={onTokenPressed(tokenInfo.address)} />
            </React.Fragment>
          )
        })}
      </Animated.ScrollView>
    </View>
  )
}

TokenBottomSheet.navigationOptions = {}

const styles = StyleSheet.create({
  container: {
    position: 'absolute',
    top: 0,
    bottom: 0,
    left: 0,
    right: 0,
    justifyContent: 'flex-end',
    zIndex: 1,
  },
  background: {
    position: 'absolute',
    backgroundColor: colors.modalBackdrop,
    opacity: 0.5,
    width: '100%',
    height: '100%',
  },
  backgroundTouchable: {
    width: '100%',
    height: '100%',
  },
  contentContainer: {
    position: 'absolute',
    bottom: 0,
    opacity: 1,
    width: '100%',
    backgroundColor: colors.light,
    padding: Spacing.Thick24,
    borderTopRightRadius: Spacing.Regular16,
    borderTopLeftRadius: Spacing.Regular16,
  },
  title: {
    ...fontStyles.h2,
    marginBottom: Spacing.Smallest8,
  },
  tokenOptionContainer: {
    flexDirection: 'row',
    alignItems: 'center',
    paddingVertical: 16,
  },
  tokenImage: {
    width: 40,
    height: 40,
    borderRadius: 20,
    marginRight: 10,
  },
  tokenNameContainer: {
    alignItems: 'flex-start',
  },
  tokenBalanceContainer: {
    flex: 1,
    alignItems: 'flex-end',
  },
  localBalance: {
    ...fontStyles.regular,
  },
  currencyBalance: {
    ...fontStyles.small,
    color: colors.gray4,
  },
  separator: {
    width: '100%',
    height: 1,
    backgroundColor: colors.gray2,
  },
  fullHeightScrollView: {
    paddingBottom: 50,
  },
})

export default TokenBottomSheet
