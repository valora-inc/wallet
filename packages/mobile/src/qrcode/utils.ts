import * as RNFS from 'react-native-fs'
import Share from 'react-native-share'
import { call, put, select } from 'redux-saga/effects'
import { showError, showMessage } from 'src/alert/actions'
import { SendEvents } from 'src/analytics/Events'
import { SendOrigin, WalletConnectPairingOrigin } from 'src/analytics/types'
import ValoraAnalytics from 'src/analytics/ValoraAnalytics'
import { ErrorMessages } from 'src/app/ErrorMessages'
import { walletConnectEnabledSelector } from 'src/app/selectors'
import { validateRecipientAddressSuccess } from 'src/identity/actions'
import { E164NumberToAddressType } from 'src/identity/reducer'
import { navigate } from 'src/navigator/NavigationService'
import { Screens } from 'src/navigator/Screens'
import { UriData, uriDataFromUrl } from 'src/qrcode/schema'
import {
  getRecipientFromAddress,
  recipientHasNumber,
  RecipientInfo,
} from 'src/recipients/recipient'
import { QrCode, SVG } from 'src/send/actions'
import { TransactionDataInput } from 'src/send/SendAmount'
import { handleSendPaymentData } from 'src/send/utils'
import Logger from 'src/utils/Logger'
import { initialiseWalletConnect } from 'src/walletConnect/saga'

export enum BarcodeTypes {
  QR_CODE = 'QR_CODE',
}

const TAG = 'QR/utils'

const QRFileName = '/celo-qr.png'

export async function shareSVGImage(svg: SVG) {
  if (!svg) {
    return
  }
  const data = await new Promise<string>((resolve, reject) => {
    svg.toDataURL((dataURL: string | undefined) => {
      if (dataURL) {
        resolve(dataURL)
      } else {
        // Not supposed to happen, but throw in case it does :)
        reject(new Error('Got invalid SVG data'))
      }
    })
  })

  const path = RNFS.DocumentDirectoryPath + QRFileName
  await RNFS.writeFile(path, data, 'base64')
  return Share.open({
    url: 'file://' + path,
    type: 'image/png',
    failOnCancel: false, // don't throw if user cancels share
  })
}

function* handleSecureSend(
  address: string,
  e164NumberToAddress: E164NumberToAddressType,
  secureSendTxData: TransactionDataInput,
  requesterAddress?: string
) {
  if (!recipientHasNumber(secureSendTxData.recipient)) {
    throw Error('Invalid recipient type for Secure Send, has no mobile number')
  }

  const userScannedAddress = address.toLowerCase()
  const { e164PhoneNumber } = secureSendTxData.recipient
  const possibleReceivingAddresses = e164NumberToAddress[e164PhoneNumber]
  // This should never happen. Secure Send is triggered when there are
  // multiple addresses for a given phone number
  if (!possibleReceivingAddresses) {
    throw Error("No addresses associated with recipient's phone number")
  }

  // Need to add the requester address to the option set in the event
  // a request is coming from an unverified account
  if (requesterAddress && !possibleReceivingAddresses.includes(requesterAddress)) {
    possibleReceivingAddresses.push(requesterAddress)
  }
  const possibleReceivingAddressesFormatted = possibleReceivingAddresses.map((addr) =>
    addr.toLowerCase()
  )
  if (!possibleReceivingAddressesFormatted.includes(userScannedAddress)) {
    const error = ErrorMessages.QR_FAILED_INVALID_RECIPIENT
    ValoraAnalytics.track(SendEvents.send_secure_incorrect, {
      confirmByScan: true,
      error,
    })
    yield put(showMessage(error))
    return false
  }

  ValoraAnalytics.track(SendEvents.send_secure_complete, { confirmByScan: true })
  yield put(validateRecipientAddressSuccess(e164PhoneNumber, userScannedAddress))
  return true
}

export function* handleBarcode(
  barcode: QrCode,
  e164NumberToAddress: E164NumberToAddressType,
  recipientInfo: RecipientInfo,
  secureSendTxData?: TransactionDataInput,
  isOutgoingPaymentRequest?: boolean,
  requesterAddress?: string
) {
  const walletConnectEnabled: boolean = yield select(walletConnectEnabledSelector)
  if (barcode.data.startsWith('wc:') && walletConnectEnabled) {
    navigate(Screens.WalletConnectLoading, { origin: WalletConnectPairingOrigin.Scan })
    yield call(initialiseWalletConnect, barcode.data, WalletConnectPairingOrigin.Scan)
    return
  }

  let qrData: UriData
  try {
    qrData = uriDataFromUrl(barcode.data)
  } catch (e) {
    yield put(showError(ErrorMessages.QR_FAILED_INVALID_ADDRESS))
    Logger.error(TAG, 'qr scan failed', e)
    return
  }

  if (secureSendTxData) {
    const success: boolean = yield call(
      handleSecureSend,
      qrData.address,
      e164NumberToAddress,
      secureSendTxData,
      requesterAddress
    )
    if (!success) {
      return
    }

    if (isOutgoingPaymentRequest) {
      navigate(Screens.PaymentRequestConfirmation, {
        transactionData: secureSendTxData,
        addressJustValidated: true,
      })
    } else {
      navigate(Screens.SendConfirmation, {
        transactionData: secureSendTxData,
        addressJustValidated: true,
        origin: SendOrigin.AppSendFlow,
      })
    }

    return
  }

  const cachedRecipient = getRecipientFromAddress(qrData.address, recipientInfo)

  yield call(handleSendPaymentData, qrData, cachedRecipient, isOutgoingPaymentRequest, true)
}
