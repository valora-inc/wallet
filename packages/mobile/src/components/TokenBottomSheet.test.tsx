import { fireEvent, render } from '@testing-library/react-native'
import * as React from 'react'
import { Provider } from 'react-redux'
import TokenBottomSheet, { TokenPickerOrigin } from 'src/components/TokenBottomSheet'
import { Currency } from 'src/utils/currencies'
import { amountFromComponent, createMockStore } from 'test/utils'
import { mockCeurAddress, mockCusdAddress } from 'test/values'

jest.mock('src/components/useShowOrHideAnimation')

const mockStore = createMockStore({
  stableToken: {
    balances: { [Currency.Dollar]: '100', [Currency.Euro]: '200' },
  },
  tokens: {
    tokenBalances: {
      [mockCusdAddress]: {
        balance: '10',
        usdPrice: '1',
        symbol: 'cUSD',
        address: mockCusdAddress,
      },
      [mockCeurAddress]: {
        balance: '20',
        usdPrice: '1.2',
        symbol: 'cEUR',
        address: mockCeurAddress,
      },
    },
  },
})

const onTokenSelectedMock = jest.fn()
const onCloseMock = jest.fn()

describe('TokenBottomSheet', () => {
  beforeAll(() => {
    // @ts-ignore This avoids an error, see: https://github.com/software-mansion/react-native-reanimated/issues/1380
    global.__reanimatedWorkletInit = jest.fn()
  })

  beforeEach(() => {
    jest.clearAllMocks()
  })

  function renderPicker(visible: boolean) {
    return render(
      <Provider store={mockStore}>
        <TokenBottomSheet
          isVisible={visible}
          origin={TokenPickerOrigin.Send}
          onTokenSelected={onTokenSelectedMock}
          onClose={onCloseMock}
        />
      </Provider>
    )
  }

  it('renders correctly', () => {
    const tree = renderPicker(true)
    const { getByTestId } = tree

    expect(tree.getByTestId('TokenBottomSheetContainer')).toBeTruthy()

    expect(amountFromComponent(getByTestId('cUSDBalance'))).toBe('10.00 cUSD')
    expect(amountFromComponent(getByTestId('LocalcUSDBalance'))).toBe('$13.30')
    expect(amountFromComponent(getByTestId('cEURBalance'))).toBe('20.00 cEUR')
    expect(amountFromComponent(getByTestId('LocalcEURBalance'))).toBe('$31.92') // 20 * 1.2 (cEUR price) * 1.33 (MXN price)

    expect(tree).toMatchSnapshot()
  })

  it('handles the choosing of a token correctly', () => {
    const { getByTestId } = renderPicker(true)

    fireEvent.press(getByTestId('cUSDTouchable'))
    expect(onTokenSelectedMock).toHaveBeenLastCalledWith(mockCusdAddress)

    fireEvent.press(getByTestId('cEURTouchable'))
    expect(onTokenSelectedMock).toHaveBeenLastCalledWith(mockCeurAddress)
  })

  it('handles taps on the background correctly', () => {
    const { getByTestId } = renderPicker(true)

    fireEvent.press(getByTestId('BackgroundTouchable'))
    expect(onCloseMock).toHaveBeenCalled()
  })

  it('renders nothing if not visible', () => {
    const { queryByTestId } = renderPicker(false)
    expect(queryByTestId('TokenBottomSheetContainer')).toBeFalsy()
  })
})
