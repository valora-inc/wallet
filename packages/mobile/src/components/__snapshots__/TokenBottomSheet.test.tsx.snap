// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TokenBottomSheet renders correctly 1`] = `
<View
  style={
    Object {
      "bottom": 0,
      "justifyContent": "flex-end",
      "left": 0,
      "position": "absolute",
      "right": 0,
      "top": 0,
      "zIndex": 1,
    }
  }
  testID="TokenBottomSheetContainer"
>
  <View
    style={
      Array [
        Object {
          "backgroundColor": "#1B392F",
          "height": "100%",
          "opacity": 0.5,
          "position": "absolute",
          "width": "100%",
        },
        undefined,
      ]
    }
  >
    <RNGestureHandlerButton
      collapsable={false}
      onGestureEvent={[Function]}
      onGestureHandlerEvent={[Function]}
      onGestureHandlerStateChange={[Function]}
      onHandlerStateChange={[Function]}
      rippleColor={0}
      testID="BackgroundTouchable"
    >
      <View
        accessible={true}
        style={
          Object {
            "height": "100%",
            "width": "100%",
          }
        }
      />
    </RNGestureHandlerButton>
  </View>
  <View
    onLayout={[Function]}
    style={
      Array [
        Object {
          "backgroundColor": "#FFFFFF",
          "borderTopLeftRadius": 16,
          "borderTopRightRadius": 16,
          "bottom": 0,
          "opacity": 1,
          "padding": 24,
          "position": "absolute",
          "width": "100%",
        },
        Object {
          "paddingBottom": 24,
        },
        undefined,
      ]
    }
  >
    <Text
      style={
        Object {
          "color": "#2E3338",
          "fontFamily": "Jost-Medium",
          "fontSize": 22,
          "lineHeight": 28,
          "marginBottom": 8,
        }
      }
    >
      selectBalance
    </Text>
    <View
      accessible={true}
      focusable={true}
      nativeBackgroundAndroid={
        Object {
          "attribute": "selectableItemBackground",
          "rippleRadius": undefined,
          "type": "ThemeAttrAndroid",
        }
      }
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      testID="cUSDTouchable"
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "paddingVertical": 16,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#2E3338",
              "flex": 1,
              "fontFamily": "Inter-Medium",
              "fontSize": 16,
              "lineHeight": 22,
            }
          }
        >
          stableBalance
        </Text>
        <View
          style={
            Object {
              "alignItems": "flex-end",
            }
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#2E3338",
                  "fontFamily": "Inter-Regular",
                  "fontSize": 16,
                  "lineHeight": 22,
                },
                Object {
                  "color": "#2E3338",
                },
              ]
            }
            testID="LocalcUSDBalance/value"
          >
            
            $
            13.30
          </Text>
          <Text
            style={
              Array [
                Object {
                  "color": "#9CA4A9",
                  "fontFamily": "Inter-Regular",
                  "fontSize": 14,
                  "lineHeight": 18,
                },
                Object {
                  "color": "#9CA4A9",
                },
              ]
            }
            testID="cUSDBalance/value"
          >
            
            10.00
             cUSD
          </Text>
        </View>
      </View>
    </View>
    <View
      style={
        Object {
          "backgroundColor": "#EDEEEF",
          "height": 1,
          "width": "100%",
        }
      }
    />
    <View
      accessible={true}
      focusable={true}
      nativeBackgroundAndroid={
        Object {
          "attribute": "selectableItemBackground",
          "rippleRadius": undefined,
          "type": "ThemeAttrAndroid",
        }
      }
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      testID="cEURTouchable"
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "paddingVertical": 16,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#2E3338",
              "flex": 1,
              "fontFamily": "Inter-Medium",
              "fontSize": 16,
              "lineHeight": 22,
            }
          }
        >
          stableBalance
        </Text>
        <View
          style={
            Object {
              "alignItems": "flex-end",
            }
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#2E3338",
                  "fontFamily": "Inter-Regular",
                  "fontSize": 16,
                  "lineHeight": 22,
                },
                Object {
                  "color": "#2E3338",
                },
              ]
            }
            testID="LocalcEURBalance/value"
          >
            
            $
            40.00
          </Text>
          <Text
            style={
              Array [
                Object {
                  "color": "#9CA4A9",
                  "fontFamily": "Inter-Regular",
                  "fontSize": 14,
                  "lineHeight": 18,
                },
                Object {
                  "color": "#9CA4A9",
                },
              ]
            }
            testID="cEURBalance/value"
          >
            
            20.00
             cEUR
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;
